File: packages/mesh-io/typescript/src/write-point-set.ts
import { PointSet, BinaryFile, getFileExtension } from "itk-wasm";

import mimeToPointSetIo from "./mime-to-point-set-io.js";
import extensionToPointSetIo from "./extension-to-point-set-io.js";
import pointSetIoIndex from "./point-set-io-index.js";
import WritePointSetOptions from "./write-point-set-options.js";
import WritePointSetResult from "./write-point-set-result.js";

interface WriterOptions {
  informationOnly?: boolean;
  useCompression?: boolean;
  /** WebWorker for computation. Set to null to create a new worker. Or, pass an existing worker. Or, set to `false` to run in the current thread / worker. */
  webWorker?: Worker | null | boolean;
}
interface WriterResult {
  webWorker: Worker;
  couldWrite: boolean;
  serializedPointSet: BinaryFile;
}
type Writer = (
  pointSet: PointSet,
  serializedPointSet: string,
  options: WriterOptions,
) => Promise<WriterResult>;

/**
 * Write an itk-wasm PointSet converted to an serialized point set file format
 *
 * @param {PointSet} pointSet - Input point set
 * @param {string} serializedPointSet - Output point set serialized in the file format.
 * @param {WritePointSetOptions} options - options object
 *
 * @returns {Promise<WritePointSetResult>} - result object
 */
async function writePointSet(
  pointSet: PointSet,
  serializedPointSet: string,
  options: WritePointSetOptions = {},
): Promise<WritePointSetResult> {
  let inputPointSet = pointSet;

  const mimeType = options.mimeType;
  const extension = getFileExtension(serializedPointSet).toLowerCase();
  let usedWebWorker = options.webWorker;

  let io = null;
  if (typeof mimeType !== "undefined" && mimeToPointSetIo.has(mimeType)) {
    io = mimeToPointSetIo.get(mimeType);
  } else if (extensionToPointSetIo.has(extension)) {
    io = extensionToPointSetIo.get(extension);
  } else {
    for (const readerWriter of pointSetIoIndex.values()) {
      if (readerWriter[1] !== null) {
        let {
          webWorker: testWebWorker,
          couldWrite,
          serializedPointSet: serializedPointSetBuffer,
        } = await (readerWriter[1] as unknown as Writer)(
          inputPointSet,
          serializedPointSet,
          options,
        );
        usedWebWorker = testWebWorker;
        if (couldWrite) {
          return {
            webWorker: usedWebWorker as Worker,
            serializedPointSet: serializedPointSetBuffer,
          };
        }
      }
    }
  }
  if (!io) {
    throw Error("Could not find IO for: " + serializedPointSet);
  }
  const readerWriter = pointSetIoIndex.get(io as string);

  const writer = (readerWriter as Array<Writer>)[1];
  let {
    webWorker: testWebWorker,
    couldWrite,
    serializedPointSet: serializedPointSetBuffer,
  } = await writer(inputPointSet, serializedPointSet, options);
  usedWebWorker = testWebWorker;
  if (!couldWrite) {
    throw Error("Could not write: " + serializedPointSet);
  }

  const result = {
    webWorker: usedWebWorker as Worker,
    serializedPointSet: serializedPointSetBuffer,
  };
  return result;
}

export default writePointSet;

File: packages/mesh-io/python/itkwasm-mesh-io-wasi/itkwasm_mesh_io_wasi/write_point_set.py
import os
import importlib
from pathlib import Path

from itkwasm import PointSet

from .extension_to_point_set_io import extension_to_point_set_io
from .point_set_io_index import point_set_io_index

def write_point_set(
    point_set: PointSet,
    serialized_point_set: os.PathLike,
    information_only: bool = False,
    use_compression: bool = False,
) -> None:
    """Write an ITK-Wasm PointSet to a point set file format.

    :param point_set: Input point set
    :type  point_set: PointSet

    :param serialized_point_set: Output point set serialized in the file format.
    :type  serialized_point_set: str

    :param information_only: Only write point set metadata -- do not write pixel data.
    :type  information_only: bool

    :param use_compression: Use compression in the written file
    :type  use_compression: bool

    :param serialized_point_set: Input point set serialized in the file format
    :type  serialized_point_set: os.PathLike
    """
    extension = ''.join(Path(serialized_point_set).suffixes)

    io = None
    if extension in extension_to_point_set_io:
        func = f"{extension_to_point_set_io[extension]}_write_point_set"
        mod_name = f"itkwasm_mesh_io_wasi.{func}"
        mod = importlib.import_module(mod_name)
        io = getattr(mod, func)
    else:
        for ioname in point_set_io_index:
            func = f"{ioname}_write_point_set"
            mod_name = f"itkwasm_mesh_io_wasi.{func}"
            mod = importlib.import_module(mod_name)
            io = getattr(mod, func)
            could_write = io(point_set, serialized_point_set, information_only=information_only, use_compression=use_compression)
            if could_write:
                return

    if io is None:
        raise RuntimeError(f"Could not find an point_set writer for {extension}")

    could_write = io(point_set, serialized_point_set, information_only=information_only, use_compression=use_compression)
    if not could_write:
        raise RuntimeError(f"Could not write {serialized_point_set}")

def pointsetwrite(
    point_set: PointSet,
    serialized_point_set: os.PathLike,
    information_only: bool = False,
    use_compression: bool = False,
) -> None:
    return write_point_set(point_set, serialized_point_set, information_only=information_only, use_compression=use_compression)

pointsetwrite.__doc__ = f"""{write_point_set.__doc__}
    Alias for write_point_set.
    """